test/xbgpu/test_engine.py

```diff
@@ -695,8 +668,15 @@ async def _send_data(
 
                 corrprod_results[i][j] = ig_recv["xeng_raw"].value
 
+        # NOTE: Update `batch_indices` to be end on a multiple of
+        # `HEAPS_PER_FENGINE_PER_CHUNK`, but only for the beam_outputs because
+        # they currently send `HEAPS_PER_FENGINE_PER_CHUNK` heaps all the time.
+        # This does not mean the final heap (for each beam_output) has sane
+        # data in it. In fact, ensure you verify data up to `n_beam_heaps - 1`.
+        n_beam_heaps = roundup(len(batch_indices), HEAPS_PER_FENGINE_PER_CHUNK)
+        beam_batch_indices = range(batch_indices[0], batch_indices[0] + n_beam_heaps)
```

Comment:
I think this logic only works if `batch_indices` is contiguous AND starts on a multiple of HEAPS_PER_FENGINE_PER_CHUNK. I don't think it's worth trying to handle the general case in this PR, although it should does need to be tested under NGC-1172. Just add a comment to document the assumption, and maybe an assertion too so that the computer will remind you when it is violated.

Reply:
Once again, too right. I completely glossed over the mention in the parameter's docstring.

Will update accordingly.